### PR TITLE
use containerd/reference/docker instead of docker/distribution

### PR DESCRIPTION
**Pull request by thaJeztah**
splitting this off from https://github.com/moby/moby/pull/40961

similar PR can be found in https://github.com/moby/buildkit/pull/2666, which removes the dependency on this package from BuildKit (only remaining use in the code now)

containerd now maintains a copy of the docker/distribution code under the
github.com/containerd/containerd/reference/docker package.

Let's switch to using that package, to remove the (direct) dependency on
github.com/docker/distribution.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
Ah, some fixes needed in a mock;

```
distribution/push_v2_test.go:404:4: cannot use repo (type *mockRepo) as type distribution.Repository in field value:
    *mockRepo does not implement distribution.Repository (wrong type for Named method)
        have Named() docker.Named
        want Named() "github.com/docker/docker/vendor/github.com/docker/distribution/reference".Named
distribution/push_v2_test.go:607:3: cannot use repo (type *mockRepoWithBlob) as type distribution.Repository in field value:
    *mockRepoWithBlob does not implement distribution.Repository (wrong type for Named method)
        have Named() docker.Named
        want Named() "github.com/docker/docker/vendor/github.com/docker/distribution/reference".Named
distribution/push_v2_test.go:638:5: cannot use &mockRepo{} (type *mockRepo) as type distribution.Repository in assignment:
    *mockRepo does not implement distribution.Repository (wrong type for Named method)
        have Named() docker.Named
        want Named() "github.com/docker/docker/vendor/github.com/docker/distribution/reference".Named
```

**Comment by thaJeztah** [comment]
So, the error is because the `mockRepo`s do not implement this interface; https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/vendor/github.com/docker/distribution/registry.go#L96-L99


But I feel silly now; not sure I understand; so the failure is because I swapped the imports. Now, that makes perfect sense if the returned type was a "type", but it's an _interface_, so this line:

https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/distribution/push_v2_test.go#L640-L643


- Previously returned `github.com/docker/distribution/reference.Named`
- But now returns `github.com/containerd/containerd/reference/docker.Named`

So.. how do they differ? Well.. they _don't_. The containerd `reference/docker` package is a copy of distribution's `reference` package;

Here's the docker/distribution definition:

https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/vendor/github.com/docker/distribution/reference/reference.go#L63-L68

https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/vendor/github.com/docker/distribution/reference/reference.go#L108-L112

And here's the containerd definition:

https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/vendor/github.com/containerd/containerd/reference/docker/reference.go#L81-L86

https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/vendor/github.com/containerd/containerd/reference/docker/reference.go#L126-L130


So, I would expect things to fail if those definitions depended on some _type_, but they're defining an interface that returns a generic (`string`) type.

File: distribution/push_v2_test.go [hunk 1]
```diff
@@ -637,7 +638,15 @@ type mockRepo struct {
...
-func (m *mockRepo) Named() reference.Named {
+// Named() must return a distref.Named here, not containerd's Named interface;
+// while both interfaces are identical, Golang considers it an incompatible
+// type otherwise:
+//
+//     distribution/push_v2_test.go:404:4: cannot use repo (type *mockRepo) as type distribution.Repository in field value:
+//     *mockRepo does not implement distribution.Repository (wrong type for Named method)
+//     have Named() docker.Named
+//     want Named() "github.com/docker/docker/vendor/github.com/docker/distribution/reference".Named
+func (m *mockRepo) Named() distref.Named {
```
Comment (thaJeztah) [inline distribution/push_v2_test.go:649]: As it looks like the "interface" issues really only affects this Mock, I decided to import the "old" package here. "production" code doesn't need the old interface

Not sure if there's a better solution for this; perhaps we should replace the `distribution.Repository` interface with something local if possible (haven't dug too deeply into that yet)

**Comment by tianon** [comment]
Relevant to what we discussed in the maintainers meeting, it appears this was copied into containerd specifically for https://github.com/containerd/containerd/issues/3554 (via https://github.com/containerd/containerd/pull/3728).

Was there ever any communication with either party about maintaining it in a separate (smaller) place?  Now there are two copies that have diverged slightly and it'd be neat if we could help get them back in sync. :see_no_evil:

**Comment by thaJeztah** [comment]
- superseded by https://github.com/moby/moby/pull/46376

**Pull request closed without merging**